src/darsia/utils/linalg.py

```diff
@@ -19,3 +19,255 @@ def __init__(self, A: sps.csc_matrix) -> None:
 
     def solve(self, b: np.ndarray, **kwargs) -> np.ndarray:
         return gmres(self.A, b, **kwargs)[0]
+
+
+def _make_reasons(reasons):
+    return dict([(getattr(reasons, r), r)
+                 for r in dir(reasons) if not r.startswith("_")])
+KSPreasons = _make_reasons(PETSc.KSP.ConvergedReason())
+
+class KSP:
+    def __init__(self, A: sps.csc_matrix, 
+                 field_ises=None, 
+                 nullspace=None,
+                 appctx: dict=None
+                 ) -> None:
+        
+        # convert csc to csr (if needed)
+        if not sps.isspmatrix_csr(A):
+            A = A.tocsr()
+
+        # store original matrix
+        self.A = A
+
+        # convert to petsc matrix
+        self.A_petsc = PETSc.Mat().createAIJ(
+            size=A.shape, csr=(A.indptr, A.indices, A.data)
+        )
+
+        # This step is needed to avoid a petsc error with LU factorization
+        # that explicitly needs a zeros along the diagonal
+        # TODO: use zero_diagonal = PETSc.Mat().createConstantDiagonal(self.A.size,0.0)
+        zero_diagonal = PETSc.Mat().createAIJ(
+            size=A.shape, 
+            csr=(np.arange(A.shape[0]+1,dtype="int32"),
+                np.arange(A.shape[0],dtype="int32"),
+                np.zeros(A.shape[0]))
+        )
+        self.A_petsc += zero_diagonal
+
+        # set nullspace
+        self._petsc_kernels = []
+        if nullspace is not None:
+            # convert to petsc vectors
+            self._comm = PETSc.COMM_WORLD
+            for v in nullspace:
+                p_vec = PETSc.Vec().createWithArray(v, comm=self._comm)
+                self._petsc_kernels.append(p_vec)
+            self._nullspace = PETSc.NullSpace().create(constant=None,
+                                                   vectors=self._petsc_kernels,
+                                                   comm=self._comm)
+            self.A_petsc.setNullSpace(self._nullspace)
+
+        # preallocate petsc vectors
+        self.sol_petsc = self.A_petsc.createVecLeft()
+        self.rhs_petsc = self.A_petsc.createVecRight()
+
+        # set field_ises
+        if field_ises is not None:
+            self.field_ises = field_ises
+        else:
+            self.field_ises = None
+
+        self.appctx = appctx
+        
+        
+
+    def setup(self, petsc_options: dict) -> None:
+        petsc_options = flatten_parameters(petsc_options)
+
+
+        self.prefix = "petsc_solver_"
```

Comment:
Does this has implications for how to define the options of the solver from the outside?